[wasm-mt] Add MessageChannel between browser thread and new pthreads

lambdageek opened the pull request:
Add a mechanism for creating a communication channel between the browser thread and new pthreads (running in webworkers - note, Emscripten may recycle workers after a pthread exits).

This is done by adding our own event listener to the webworker (in the main thread) and to globalThis (in the worker).  This conflicts with emscripten's message handlers (although it's considered a bug by Emscripten upstream that their event handler doesn't ignore unrecognized messages).

One potential problem here is that for any communication to happen, the worker must service its event loop.  If it's just busy running a loop in wasm, it might never handle the messages.  On the other hand, posting messages back to main should work.  

Once we have our message handlers in place, the rest is straightforward, the worker creates a MessageChannel and transfers one of the ports to the browser thread.

With the browser-to-pthread channel established, we can build up cross-thread channels by asking the main thread to transfer ports on our behalf.  This part isn't done yet.

lambdageek committed:
outline of a dedicated channel between JS main thread and pthreads

lambdageek committed:
WIP: wire up the MessageChannel to native thread attach

lambdageek committed:
add JS entrypoints for pthread-channel

lambdageek committed:
debug printfs etc

lambdageek committed:
more printfs

lambdageek committed:
comment out the thread start.

was used to validate that a pthread message channel is working.

Not needed in this sample, because the internal EventPipe sampling
thread is good enough as proof

lambdageek commented:
Also I'm not happy with everything lumped in one .ts file.  Next push to this draft will factor the code into stuff that the browser runs, stuff that the workers run, and shared code.  (ultimately everything here is rolled up into dotnet.js, but it's easier to think about if the code is in separate modules)

lambdageek committed:
split up into pthread-channel module into worker, browser and shared

lambdageek committed:
WIP: pthreads modules

lambdageek committed:
add ENVIRONMENT_IS_PTHREAD; mono_wasm_pthread_worker_init

lambdageek committed:
Fixup names; call MessagePort.start; remove printfs

lambdageek committed:
remove whitespace and extra printfs

pavelsavara commented on src/mono/wasm/runtime/pthreads/README.md:
```diff
@@ -0,0 +1,50 @@
+
+  This could get better if the following things changed in Emscripten:
+
+  1. There was some hook on Module.Pthread.loadWasmModuleToWorker so that we could add our own message handler at the same time that Emscripten adds its own.
```
We already replace some implementations of emscripten functions to hook into it. See `__dotnet_replacements.updateGlobalBufferAndViews`

lambdageek replied:
Done.  This simplified things a lot, thanks!

pavelsavara commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -70,6 +70,8 @@ import { EmscriptenModule } from "./types/emscripten";
 import { mono_run_main, mono_run_main_and_exit } from "./run";
 import { diagnostics } from "./diagnostics";
 import { dotnet_browser_can_use_simple_digest_hash, dotnet_browser_simple_digest_hash } from "./crypto-worker";
+import { mono_wasm_pthread_on_pthread_created } from "./pthreads/worker";
```
This would get included even in non-threaded runtime. Maybe we could do something with `rollup` to link it out. See `rollup.config.js` -> `productVersion` and also `rollup-plugin-consts`, I think it could eliminate dead code on constants.

lambdageek replied:
I think I can almost make this work, but how do I make one of these rollup constants visible in `dotnet.cjs.lib.js`? I need to conditionally fill `linked_functions`.

pavelsavara replied:
Create `dotnet.cjs.mt.lib.js` and do it in build files ?

We need to do something smarter about those `linked_functions` in the future. 
I'm not confident what we have is the best solution.
There is some crazy transormation on it by emscripten.

lambdageek replied:
oh, I guess the lib.js files go through emcc's weird C preprocessor JS thing, so I can do something with `#ifdef`.

lambdageek replied:
oh god, it's not the real C preprocessor. it's their own homegrown thing...

https://github.com/emscripten-core/emscripten/blob/54d1aa0917e91c60533dd0b6bee2212d487cc78a/src/parseTools.js#L37

lambdageek replied:
Ok, pushed a fix for this.  verified that a single-threaded Release build of `dotnet.js` doesn't have any of the pthread support code in it

lambdageek committed:
Exclude threading exports in non-threaded runtime

Use the `USE_THREADS` emscripten library ifdef to prevent the
entrypoints from being saved.

Use a new `MonoWasmThreads` rollup constant to remove JS roots.

Verified that a Release build single-threaded dotnet.js doesn't
include any of the new pthread support code

lambdageek committed:
Add replacement for PThread.loadWasmModuleToWorker

This will allow us to install a message handler when the worker is
created, before it has any pthreads assigned to it.

We can therefore simplify how we set up our own MessageChannel to
simply send a single event from the worker to the browser thread,
instead of having to lazily install the event handler on the main
thread by queueing async work to the browser thread.

lambdageek committed:
Simplify the dedicated channel creation

now that we can add a message handler to a worker when Emscripten
creates it, skip the complicated atomic notification process

lambdageek commented on src/mono/mono/utils/mono-threads-wasm.c:
```diff
@@ -411,6 +411,27 @@ mono_threads_platform_stw_defer_initial_suspend (MonoThreadInfo *info)
+	}
+	// Set up a MessageChannel between the new thread (which might be on a pooled reused WebWorker) and the main thread.
+	pthread_t id = pthread_self ();
+	mono_wasm_pthread_on_pthread_created (id);
```
```suggestion
	MONO_ENTER_GC_SAFE;
	mono_wasm_pthread_on_pthread_created (id);
	MONO_EXIT_GC_SAFE;
```

Hmm. I probably should go back through my earlier PRs and make sure I'm doing a GC Safe transition on exits from C to JS...

lambdageek committed:
Don't forget the GC transition out to JS

lambdageek committed:
fixup rebase

lambdageek commented:
I think this one is ready for review, @pavelsavara @kg. I'm working on the next set of changes (primarily around starting a diagnostic server), and there doesn't seem to be too much else that I want to add to the infrastructure here

pavelsavara commented on src/mono/wasm/runtime/cjs/dotnet.cjs.lib.js:
```diff
@@ -4,22 +4,34 @@
 "use strict";
 
+#if USE_PTHREADS
+const usePThreads = `true`;
```
do you really want string and not bool ?

pavelsavara replied:
Which tool does the interpolation here? If that was JS it would be \`${true}\`

lambdageek replied:
It goes through Emscripten's preprocessor (which takes care of the `#if`), and then `$DOTNET__postset` this whole file runs during emscripten link time. So the `usePThreads` string is interpolated into the `$DOTNET__postset` string (so we get `$DOTNET__postset = "let __dotnet_replacement_PThread_loadWasmModuleToWorker = ${usePThreads} ? PThread.loadWasmModuleToWorker : null..."`.  Then the `$DOTNET__postset` string gets used by the emscripten library manager and gets pasted into the body of the emscripten IIFE.

The upshot is that if we're in `#if USE_PTHREADS`, the body of the emscripten module will contain `let __dotnet_replacement_PThread_loadWasmModuleToWorker = false ? PThread.loadWasmModuleToWorker : undefined;`.  And then rollup collapses that to just `let __dotnet_replacement_PThread_loadWasmModuleToWorker  = undefined` which is what we want when there's no threads.

pavelsavara replied:
This file is not processed by rollup at all. Maybe closure compiler in emcc does that ?
I understand that the preprocesor would handle `#if`. I don't follow that it would take the code between `#if` and `#else` and evaluate it. I think it would just keep is as is as text, no ?

pavelsavara replied:
I get it now, the evaluation of `DOTNET__postset` would do the `${isPThread}` evaluation just before it enters closure compiler.

pavelsavara commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -376,7 +387,12 @@ export const __linker_exports: any = {
+    dotnet_browser_sign,
+
+    // mono-threads-wasm.c
+    ...(MonoWasmThreads) && {
```
are `...` also unpacking the object after `&&` ?

lambdageek replied:
StackOverflow told me this is how you make ... either unpack the `{ }` object after the `&&` or expand to nothing if `MonoWasmThreads` is falsy.

lambdageek replied:
https://jsfiddle.net/65hLcotp/6/

![Screen Shot 2022-06-22 at 10 23 39](https://user-images.githubusercontent.com/480437/175053463-7cb565ad-dec8-4872-937f-14437da8229d.png)
![Screen Shot 2022-06-22 at 10 23 50](https://user-images.githubusercontent.com/480437/175053465-34d8b35e-d660-4e96-a416-10373a169438.png)

pavelsavara replied:
Okay, more readable would be. 

```
let x = {a:"a",b:"b",c:"c"} ;
if (flag){
    x = {
        ...x, 
        d:"d"
    };
}

```

No so cool tho ;)

lambdageek replied:
but then it's not `const` :cry: 

I'll clean it up.

lambdageek committed:
less code golf

pavelsavara commented on src/mono/wasm/runtime/pthreads/browser/index.ts:
```diff
@@ -0,0 +1,87 @@
+}
+
+function removeThread(pthread_ptr: pthread_ptr): void {
+    threads.delete(pthread_ptr);
```
Should we send message to other workers, to unregister the thread id?

pavelsavara commented on src/mono/wasm/runtime/pthreads/worker/index.ts:
```diff
@@ -0,0 +1,34 @@
+
+export type ThreadCreatedCallback = (pthread_ptr: pthread_ptr, main_port: MessagePort) => void;
+
+const threadCreatedCallbacks = Array<ThreadCreatedCallback>();
```
This collection is in each worker ?
Does it contain list of all other workers ?
What is it good for ?

What makes it `callback` ?

lambdageek replied:
This is a generic callback to do something every time that Emscripten creates a pthread inside of a worker - before the pthread gets attached to mono inside the runtime.

The list of callbacks is in dotnet.js, so each worker will have its own list (but in practice this is going to be the same list in every thread since they will just register callbacks just after the wasm module is loaded inside the worker.

pavelsavara replied:
The later instances would not have others which registered earlier than they existed, right ?

pavelsavara commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -747,3 +753,17 @@ export type DownloadAssetsContext = {
+async function mono_wasm_pthread_worker_init(): Promise<void> {
+    // This is a good place to attach listeners for the pthread MessagePorts using pthreads_worker.addThreadCreatedCallback();
+    console.debug("mono_wasm_pthread_worker_init");
+    pthreads_worker.addThreadCreatedCallback((thread_id) => {
```
This is confusing me, it seems that this is called on the worker, right ?
Bu the callback collection is not shared across workers ?

lambdageek replied:
every time dotnet.js is loaded, the callback array is initialized.  if you load dotnet.js in two different webworkers, you will get two different arrays.  there is no shared memory (except SharedArrayBuffer) in WebWorkers.  each one is completley isolated.

pavelsavara commented:
Could we have simple sample which would play some ping pong across at least 3 threads ? 
To prove that they to talk to each other any to any.
To show how third is added and then removed and the other two could still talk.

pavelsavara commented:
If the worker is running the `startup.ts` how does it get the config we put into `createDotnetRuntime` call ?

lambdageek commented:
> Could we have simple sample which would play some ping pong across at least 3 threads ? To prove that they to talk to each other any to any. To show how third is added and then removed and the other two could still talk.

Not with this PR. This doesn't set up a way for any two workers to talk to each other. Each pthread can only talk to the main thread.

There is a limitation here, also - to be able to receive messages, each pthread would need to run the JS event loop in its worker.  That is done by calling `exit_with_live_runtime` and then returning from the pthread start function.  Most of our threads don't do this (yet - I think we might need to do it, actually, and then manually call `pthread_exit` to make the thread joinable when we're done servicing async callbacks).  See https://github.com/emscripten-core/emscripten/issues/13193 and https://github.com/emscripten-core/emscripten/pull/13596

lambdageek commented:
> If the worker is running the `startup.ts` how does it get the config we put into `createDotnetRuntime` call ?

You mean in something like `main.js`?  good question. I don't think it does.  This is what `dotnet.worker.js` does:

```js
var Module = {}

self.onmessage = (e) => {
    ...
    if (e.data.cmd === 'load') { // Preload command that is called once per worker to parse and load the Emscripten code.

      // Module and memory were sent from main thread
      Module['wasmModule'] = e.data.wasmModule;

      Module['wasmMemory'] = e.data.wasmMemory;

      Module['buffer'] = Module['wasmMemory'].buffer;

      Module['ENVIRONMENT_IS_PTHREAD'] = true;

      if (typeof e.data.urlOrBlob == 'string') {
        importScripts(e.data.urlOrBlob);
      } else {
        var objectUrl = URL.createObjectURL(e.data.urlOrBlob);
        importScripts(objectUrl);
        URL.revokeObjectURL(objectUrl);
      }
      createDotnetRuntime(Module).then(function (instance) {
        Module = instance;
      });
  } ...
  ...
}
```

If we want to do more, then we probably need our own `dotnet.worker.js` that sets up additional properties on `Module`

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-mt-channels

lambdageek committed:
fix browser-eventpipe default import

lambdageek committed:
move mono_threads_wasm_on_thread_attached later in register_thread

Actually attach the thread to the runtime (so that GC transitions
work) before calling out to JS

lambdageek commented:
Merged with main after the ES6 PR. verified that browser-eventpipe works with ES modules (needed a small fixup in main.js)

lambdageek committed:
also fix default import in browser-mt-eventpipe

lambdageek committed:
Add replacement for Module.PThread.threadInit

Use it to call mono_wasm_pthread_on_pthread_created

Rename the previous callback to mono_wasm_pthread_on_pthread_attached
- it gets called but it's unused.

This is enough to get the diagnostic sever worker started up and
pinging.

lambdageek committed:
Cleanup mono_wasm_pthread_on_pthread_created

lambdageek committed:
fix comment

lambdageek commented:
Cleaned up a little bit: I now have a "on created" and "on attached" calls - the first is called from Emscripten's `PThread["threadInit"]` function which initializes a worker once a pthread is assgned to it; the latter "on attached" is called when the thread beings to interact with the Mono runtime.  This means we can now notice when third party libraries call `pthread_create` (which might turn out ot be useful).  Also the diagnostic server thread can now be started before the native mono runtime is initialized sufficiently to create threads.

lambdageek committed:
Share tsconfig parts using "extends" property

pavelsavara commented:
could we perhaps have something like pattern [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) named `onThreadCreated`.

And subscribers would call `onThreadCreated.addEventListener(event=>{ mysomething(event.pthread_ptr) })`
and the emscripten would call `onThreadCreated.dispatchEvent()`

lambdageek committed:
Use an EventTarget and custom events for worker thread lifecycle notifications

lambdageek commented:
done.

subsystems in the runtime can now do:

```typescript
import { currentWorkerThreadEvents, dotnetPthreadCreated, WorkerThreadEvent } from './pthreads/worker';

currentWorkerThreadEvents.addEventListener(dotnetPthreadCreated, (ev: WorkerThreadEvent) => {
  /* do stuff with ev.pthread_ptr */
});
```

lambdageek committed:
pass portToMain in ThreadEvents; update README

this lets pthread lifecycle event handlers post messages and setup
listeners on the message port back to the main browser thread.

Also update the README to describe the current design

radical commented:
[Trimming tests](https://dev.azure.com/dnceng/public/_build/results?buildId=1852676&view=logs&jobId=5109813b-4d63-5536-a11d-4f34d49b9d0e&j=5109813b-4d63-5536-a11d-4f34d49b9d0e&t=46140a4e-1004-5dc2-1c6b-86b28efda89c) failing:
```
  console.error: failed to load the dotnet.js file.
  ReferenceError: CustomEvent is not defined
      at /__w/1/s/artifacts/bin/trimmingTests/projects/System.Data.Common.TrimmingTests/CreateSqlXmlReader/browser-wasm/bin/Release/net7.0/browser-wasm/AppBundle/dotnet.js:3:75595
      at /__w/1/s/artifacts/bin/trimmingTests/projects/System.Data.Common.TrimmingTests/CreateSqlXmlReader/browser-wasm/bin/Release/net7.0/browser-wasm/AppBundle/dotnet.js:3:112494
```

lambdageek commented:
> [Trimming tests](https://dev.azure.com/dnceng/public/_build/results?buildId=1852676&view=logs&jobId=5109813b-4d63-5536-a11d-4f34d49b9d0e&j=5109813b-4d63-5536-a11d-4f34d49b9d0e&t=46140a4e-1004-5dc2-1c6b-86b28efda89c) failing:
> 
> ```
>   console.error: failed to load the dotnet.js file.
>   ReferenceError: CustomEvent is not defined
>       at /__w/1/s/artifacts/bin/trimmingTests/projects/System.Data.Common.TrimmingTests/CreateSqlXmlReader/browser-wasm/bin/Release/net7.0/browser-wasm/AppBundle/dotnet.js:3:75595
>       at /__w/1/s/artifacts/bin/trimmingTests/projects/System.Data.Common.TrimmingTests/CreateSqlXmlReader/browser-wasm/bin/Release/net7.0/browser-wasm/AppBundle/dotnet.js:3:112494
> ```

Yea Im' working on it.

There's two issues:

1. `WorkerThreadEvent` isn't getting removed from builds without threads.  I'm working on making rollup remove it.
2. Node doesn't have `CustomEvent`.  Hopefully (1) will fix it, but I might just switch to `Event` instead which should be in recent Nodes

lambdageek committed:
make pthread/worker/events friendlier to tree shaking and node

pavelsavara commented:
LGTM

pavelsavara commented on src/mono/wasm/runtime/pthreads/worker/events.ts:
```diff
@@ -0,0 +1,60 @@
+    addEventListener(type: string, callback: EventListenerOrEventListenerObject | null, options?: boolean | AddEventListenerOptions): void;
+}
+
+// export class WorkerThreadEventTargetImpl extends EventTarget implements WorkerThreadEventTarget {
```
could be removed now?

lambdageek committed:
remove comment out old code

lambdageek committed:
another approach to tree shaking

rollup doesn't understand fallthru.  In the following (which is what
`mono_assert` ammounts to) it retains `C`:

```
if (condition)
  throw new Error (...);
// fallthru
return new C();
```

Solution is to use an if-then-else

```
if (condition)
  throw new Error (...);
else
  return new C(); // C is not retained if 'condition' is false
```

lambdageek commented:
weird, apparently rollup doesn't understand that if `condition` is `false` then in the following it doesn't need to retain `C`

```typescript
import condition from "consts:condition";

class C {
  constructor () {}
  ...
}

export function makeC () {
  if (!condition)
    throw new Error ("condition is false");
  // fallthru
  return new C(); // rollup incorrectly retains C
}
```

It rewrites the function to always throw, but still keeps the definition of `C`:
```typescript
class C {
  constructor() { }
  ...
}
export function makeC() {
   throw new Error ("condition is false");
}
```

However this works:

```typescript
import condition from "consts:condition";

class C {
  constructor () {}
  ...
}

export function makeC () {
  if (!condition)
    throw new Error ("condition is false");
  else
     return new C(); // rollup doesn't retain C
}
```

lambdageek committed:
more treeshaking

lambdageek committed:
fix annoying VSCode ESLint toast

Cannot read property 'loc' of undefined

See https://github.com/microsoft/vscode-eslint/issues/1149

and a proposed workaround in https://github.com/eslint/eslint/issues/14538#issuecomment-862280037

lambdageek committed:
Add Event and EventTarget polyfill for v8

lambdageek commented:
v8 shell doesn't have `Event` and `EventTarget` as far as I can tell.  Not sure how much we care about using threads with v8, but I added some poor polyfills in case we need them.

lambdageek commented on src/mono/mono/utils/mono-threads-wasm.c:
```diff
@@ -411,6 +411,29 @@ mono_threads_platform_stw_defer_initial_suspend (MonoThreadInfo *info)
+	if (mono_threads_wasm_is_browser_thread ()) {
+		return;
+	}
+	// Set up a MessageChannel between the new thread (which might be on a pooled reused WebWorker) and the main thread.
```
```suggestion
	// Notify JS that the pthread attachd to Mono
```

lambdageek commented on src/mono/mono/utils/mono-threads-wasm.c:
```diff
@@ -423,8 +446,17 @@ mono_threads_wasm_async_run_in_main_thread_vi (void (*func) (gpointer), gpointer
+
 #endif /* DISABLE_THREADS */
 
+
```
```suggestion
```

lambdageek committed:
fix whitespace and comments

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-mt-channels
